### PR TITLE
Add Atomic::fetch_update

### PR DESCRIPTION
**Pull request by PatrickNorton**
Equivalent of [`AtomicN::fetch_update`](https://doc.rust-lang.org/nightly/core/sync/atomic/struct.AtomicUsize.html#method.fetch_update) and [`AtomicCell::fetch_update`](https://github.com/crossbeam-rs/crossbeam/pull/704).

### CODE REVIEW

File: crossbeam-epoch/src/atomic.rs [hunk 1]
```diff
@@ -561,6 +561,84 @@ impl<T: ?Sized + Pointable> Atomic<T> {
...
+    ///
+    /// let res2 = a.fetch_update(|x, guard| None, SeqCst, SeqCst, guard);
+    /// assert!(res2.is_err());
+    /// ```
+    pub fn fetch_update<'g, F>(
+        &self,
+        mut func: F,
+        set_ord: Ordering,
+        fail_ord: Ordering,
+        guard: &'g Guard,
```
Comment (taiki-e) [inline crossbeam-epoch/src/atomic.rs:608]: Could you move the closure to the last argument? (like std's [`AtomicN::fetch_update`](https://doc.rust-lang.org/nightly/std/sync/atomic/struct.AtomicUsize.html#method.fetch_update). see https://github.com/rust-lang/rust/pull/71843#issue-412603003 for reasons)

File: crossbeam-epoch/src/atomic.rs [hunk 2]
```diff
@@ -561,6 +561,84 @@ impl<T: ?Sized + Pointable> Atomic<T> {
...
+                }
+            }
+            Err(prev)
+        }
+
+        fetch_update(&self.data, set_ord, fail_ord, |x| {
+            func(unsafe { Shared::from_usize(x) }, guard).map(Shared::into_usize)
+        })
+        .map(|old| unsafe { Shared::from_usize(old) })
+        .map_err(|current| unsafe { Shared::from_usize(current) })
```
Comment (taiki-e) [inline crossbeam-epoch/src/atomic.rs:639]: Why did you not use `Atomic::compare_exchange_weak` directly?
```rust
        let mut prev = self.load(fetch_order, guard);
        while let Some(next) = f(prev, guard) {
            match self.compare_exchange_weak(prev, next, set_order, fetch_order, guard) {
                Ok(shared) => return Ok(shared),
                Err(next_prev) => prev = next_prev.current,
            }
        }
        Err(prev)
```

File: crossbeam-epoch/src/atomic.rs [hunk 3]
```diff
@@ -561,6 +561,84 @@ impl<T: ?Sized + Pointable> Atomic<T> {
...
+    /// assert!(res1.is_ok());
+    ///
+    /// let res2 = a.fetch_update(|x, guard| None, SeqCst, SeqCst, guard);
+    /// assert!(res2.is_err());
+    /// ```
+    pub fn fetch_update<'g, F>(
+        &self,
+        mut func: F,
+        set_ord: Ordering,
+        fail_ord: Ordering,
```
Comment (taiki-e) [inline crossbeam-epoch/src/atomic.rs:607]: I feel that the variable names used in std [`AtomicN::fetch_update`](https://doc.rust-lang.org/nightly/std/sync/atomic/struct.AtomicUsize.html#method.fetch_update) are clearer.
```suggestion
        set_order: Ordering,
        fetch_order: Ordering,
```

**Review by taiki-e — COMMENTED** [review]
Thanks!

File: crossbeam-epoch/src/atomic.rs [hunk 4]
```diff
@@ -561,6 +561,65 @@ impl<T: ?Sized + Pointable> Atomic<T> {
...
+    /// ```
+    pub fn fetch_update<'g, F>(
+        &self,
+        set_order: Ordering,
+        fail_order: Ordering,
+        guard: &'g Guard,
+        mut func: F,
+    ) -> Result<Shared<'g, T>, Shared<'g, T>>
+    where
+        F: for<'a> FnMut(Shared<'a, T>, &'a Guard) -> Option<Shared<'a, T>>,
```
Comment (taiki-e) [inline crossbeam-epoch/src/atomic.rs:611]: Why do you use a different lifetime here instead of `'g`?

Comment (PatrickNorton) [reply crossbeam-epoch/src/atomic.rs:611]: I'm not entirely sure anymore; thinking about it again, I think that makes more sense.

File: crossbeam-epoch/src/atomic.rs [hunk 5]
```diff
@@ -561,6 +561,65 @@ impl<T: ?Sized + Pointable> Atomic<T> {
...
+    /// ```
+    pub fn fetch_update<'g, F>(
+        &self,
+        set_order: Ordering,
+        fail_order: Ordering,
+        guard: &'g Guard,
+        mut func: F,
+    ) -> Result<Shared<'g, T>, Shared<'g, T>>
+    where
+        F: FnMut(Shared<'g, T>, &'g Guard) -> Option<Shared<'g, T>>,
```
Comment (taiki-e) [inline crossbeam-epoch/src/atomic.rs:611]: Hmm, given that `&'g Guard` is an immutable reference, I'm not sure if we need to pass it as an argument to the closure.

**Review by taiki-e — APPROVED** [review]
LGTM. Thanks @PatrickNorton!

bors r+

**Comment by PatrickNorton** [comment]
Thanks for your review help!

**Pull request merged**